### PR TITLE
openrefine: init at 3.7.7

### PR DESCRIPTION
**Pull request by risicle**
## Description of changes
Addresses #214080. This tool deserves a package, as quirky and non-ideal as it is.

Built "from source" but inevitably depends on binary bytecode from maven :(

Can't get it working properly on darwin, but I don't think I've ever got anything java-y from nix running properly on macos.

cc @hervyqa

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by das-g** [comment]
Result of `nixpkgs-review pr 265959` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>openrefine</li>
  </ul>
</details>

**Comment by das-g** [comment]
console log from ungoogled-chromium: [127.0.0.1-1706709617574.log](https://github.com/NixOS/nixpkgs/files/14112596/127.0.0.1-1706709617574.log)

console log from Firefox: [console-export-2024-1-31_15-3-28.txt](https://github.com/NixOS/nixpkgs/files/14112622/console-export-2024-1-31_15-3-28.txt)

I guess `Uncaught ReferenceError: jQuery is not defined` and the following `Uncaught TypeError: Refine is undefined` might cause the problems I'm experiencing?

**Comment by risicle** [comment]
Well yes, looks like one problem is that the contents of `lib/webapp/modules/core/3rdparty` have somehow become self-referencing symlinks :confused:

**Comment by risicle** [comment]
Fixed for me. Not sure why I was trying to link instead of copy.

**Comment by risicle** [comment]
Rebased and bumped to 3.7.7

**Comment by risicle** [comment]
Approval?

**Comment by das-g** [comment]
Works for me now, but I'm unsure whether I'm qualified to review the code.

Would you be OK with a review asking a lot of naïve questions?

File: pkgs/applications/science/misc/openrefine/default.nix [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+  nativeBuildInputs = [ makeWrapper ];
+
+  installPhase = let
+    gitProperties = writeText "git.properties" (builtins.toJSON {
+      "git.build.version" = version;
+      "git.branch" = "none";
+      "git.build.time" = "1970-01-01T00:00:00+0000";
+      "git.commit.id.abbrev" = "none";
+      "git.commit.id.describe" = "none";
+    });
```
Comment (das-g) [inline pkgs/applications/science/misc/openrefine/default.nix:73]: Is setting these to constant values _the right way™_ to achieve reproducibility, or should we patch `pom.xml` with
```xml
       <properties>
         <project.build.outputTimestamp>2023-01-01T00:00:00Z</project.build.outputTimestamp>
       </properties>
```
instead, as described in the Maven [Guide to Configuring for Reproducible Builds](https://maven.apache.org/guides/mini/guide-reproducible-builds.html#how-do-i-configure-my-maven-build)?

If it is indeed the right way, doesn't Nix or stdenv.lib or so provide `"1970-01-01T00:00:00+0000"` as a constant somewhere, considering that it's probably needed at more than one place?

File: pkgs/applications/science/misc/openrefine/default.nix [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+
+    cat >> $out/etc/refine.ini <<EOF
+    REFINE_WEBAPP='$out/lib/webapp'
+    REFINE_LIB_DIR='$out/lib/server/target/lib'
+
+    JAVA_HOME='${jdk.home}'
+
+    # non-headless mode tries to launch a browser, causing a
+    # number of purity problems
+    JAVA_OPTIONS='-Drefine.headless=true'
```
Comment (das-g) [inline pkgs/applications/science/misc/openrefine/default.nix:115]: That'd be at runtime, not build time, would it? Are we now also trying to achieve purity at runtime? Or is OpenRefine being run during the builds of any other software?

(Or does non-headless mode try to modify stuff that should at runtime be immutable?)

File: pkgs/applications/science/misc/openrefine/update.sh [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+MVNHASH_OLD=$(nix-instantiate --eval . -A "${UPDATE_NIX_ATTR_PATH}.mvnHash" | tr -d '"')
+MVNHASH_OLD_ESCAPED=$(echo "${MVNHASH_OLD}" | sed -re 's|[+]|\\&|g')
+FAKEHASH=$(nix-instantiate --eval . -A "lib.fakeHash" | tr -d '"')
+FAKEHASH_ESCAPED=$(echo "${FAKEHASH}" | sed -re 's|[+]|\\&|g')
+
+sed -E -i "s|${MVNHASH_OLD_ESCAPED}|${FAKEHASH}|g" "${FILE}"
+
+MVNHASH_NEW="$(nix-build . -A "${UPDATE_NIX_ATTR_PATH}" 2>&1 | tail -n10 | grep 'got:' | cut -d: -f2- | xargs echo || true)"
+
+sed -E -i "s|${FAKEHASH_ESCAPED}|${MVNHASH_NEW}|g" "${FILE}"
```
Comment (das-g) [inline pkgs/applications/science/misc/openrefine/update.sh:20]: Where can I read up about best practices for update scripts?

File: pkgs/applications/science/misc/openrefine/update.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+FILE="$(nix-instantiate --eval -E 'with import ./. {}; (builtins.unsafeGetAttrPos "version" '"${UPDATE_NIX_ATTR_PATH}"').file' | tr -d '"')"
+
+MVNHASH_OLD=$(nix-instantiate --eval . -A "${UPDATE_NIX_ATTR_PATH}.mvnHash" | tr -d '"')
+MVNHASH_OLD_ESCAPED=$(echo "${MVNHASH_OLD}" | sed -re 's|[+]|\\&|g')
+FAKEHASH=$(nix-instantiate --eval . -A "lib.fakeHash" | tr -d '"')
+FAKEHASH_ESCAPED=$(echo "${FAKEHASH}" | sed -re 's|[+]|\\&|g')
+
+sed -E -i "s|${MVNHASH_OLD_ESCAPED}|${FAKEHASH}|g" "${FILE}"
+
+MVNHASH_NEW="$(nix-build . -A "${UPDATE_NIX_ATTR_PATH}" 2>&1 | tail -n10 | grep 'got:' | cut -d: -f2- | xargs echo || true)"
```
Comment (das-g) [inline pkgs/applications/science/misc/openrefine/update.sh:18]: Isn't there some better way to determine the new hash than trying to build with a fake hash and grepping the actual hash out of the error message?

File: pkgs/applications/science/misc/openrefine/default.nix [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+      cp $NEW_PACKAGE_JSON package.json
+    '';
+
+    dontNpmBuild = true;
+    installPhase = ''
+      mkdir -p $out
+      (
+        cd modules/core/3rdparty
+        cp -r * $out/
+      )
```
[the remark this answers is not in the document]
Comment (risicle) [reply pkgs/applications/science/misc/openrefine/default.nix:49]: Oh yeah it probably mattered when I was doing a `ln -s` but less so now.

File: pkgs/applications/science/misc/openrefine/default.nix [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+in maven'.buildMavenPackage {
+  inherit src version;
+
+  pname = "openrefine";
+
+  postPatch = ''
+    cp -r ${npmPkg} main/webapp/modules/core/3rdparty
+  '';
+  mvnParameters = "-DskipTests=true -pl !packaging";
```
[the remark this answers is not in the document]
Comment (risicle) [reply pkgs/applications/science/misc/openrefine/default.nix:58]: From memory this is just one of those things where the build Does Weird Things without it set. Getting it building at all was at the edge of my abilities TBH.

File: pkgs/applications/science/misc/openrefine/default.nix (hunk 1 above)
Comment (risicle) [reply pkgs/applications/science/misc/openrefine/default.nix:70]: Again, from memory, the build process is weird and inspects this file in some way - this was my way of stopping it.

File: pkgs/applications/science/misc/openrefine/default.nix (hunk 2 above)
Comment (risicle) [reply pkgs/applications/science/misc/openrefine/default.nix:112]: These aren't environment variables being set here, they're the contents of `$out/etc/refine.ini` being assembled.

Here I'm just trying to make it behave by default in a more well-behaved & unsurprising manner.

File: pkgs/applications/science/misc/openrefine/update.sh (hunk 4 above)
Comment (risicle) [reply pkgs/applications/science/misc/openrefine/update.sh:18]: If there is, you'll have to show me it, and also show it to all the other authors of update scripts who use this technique ;)

I didn't invent this - look throughout nixpkgs for people grepping for `got:`

Comment (das-g) [reply pkgs/applications/science/misc/openrefine/update.sh:18]: Huh, indeed:
```
$ git ls-files '**/update.sh' | wc -l
208

$ git grep 'got:' | wc -l
54
```
So approximately one out of every four update scripts uses that. I guess the others (mostly findable by `git grep -L 'got:' '**/update.sh'`) only need hashes determinable by `nix-prefetch-url` or rely on `update-source-version` (which *also* [uses that grep-for-`got:` approach](https://github.com/NixOS/nixpkgs/blob/88e3ac79cb08011d02518df0bd29b0ad8ed5372c/pkgs/common-updater/scripts/update-source-version#L242), at least for some cases.)

**Review by das-g — APPROVED** [review]
Let's leave it at that, unless someone can suggest (concrete) improvements. (I certainly can't. :see_no_evil:)

Thank you for packaging OpenRefine!

File: pkgs/applications/science/misc/openrefine/default.nix (hunk 2 above)
Comment (das-g) [reply pkgs/applications/science/misc/openrefine/default.nix:112]: > These aren't environment variables being set here, they're the contents of `$out/etc/refine.ini` being assembled.

Yeah, I'm aware of that.

> Here I'm just trying to make it behave by default in a more well-behaved & unsurprising manner.

Understood. But does that relate to purity (which in the nix ecosystem is AFAIK usually taken to apply to _build process_ purity)?

And is not launching a browser really less surprising, when the upstream software _does_ launch one by default? (Though, the PGAdmin4 packaged in NixPkgs also doesn't launch one, not even in [its `*-desktopmode` variant](https://search.nixos.org/packages?channel=unstable&show=pgadmin4-desktopmode).)

**Comment by infinisil** [comment]
Please don't merge with CI failing. It's not a big problem in this case, but new packages should use `pkgs/by-name` now, that's what the CI failure is for :)

**Comment by das-g** [comment]
> […] new packages should use `pkgs/by-name` now

Even if the second argument to `callPackage` isn't the default `{}`? (Here it's `{ jdk = jdk17; }`.)

**Comment by infinisil** [comment]
@das-g Yup. The error message even mentions that ;)

> pkgs.openrefine: This is a new top-level package of the form `callPackage ./pkgs/applications/science/misc/openrefine { }`. Please define it in pkgs/by-name/op/openrefine/package.nix instead. See `pkgs/by-name/README.md` for more details. Since the second `callPackage` argument is not `{ }`, the manual `callPackage` (e.g. in `pkgs/top-level/all-packages.nix`) is still needed.

**Pull request merged**